File: backend/app/schemas/attendance.py
from datetime import date, datetime, time
from typing import List, Optional, Literal
from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.core.timezone import to_school_timezone
from app.models.enums import AttendanceEventType, AttendanceStatus
from app.schemas.lesson_delay import LessonDelayRead


class AttendanceScanRequest(BaseModel):
    school_id: str = Field(..., min_length=36, max_length=36, description="Мектептин IDси")
    qr_token: str = Field(..., min_length=16, max_length=128, description="Сканерленген QR токен")
    latitude: float = Field(..., ge=-90.0, le=90.0, description="GPS кеңдик")
    longitude: float = Field(..., ge=-180.0, le=180.0, description="GPS узундук")
    accuracy: float = Field(..., ge=0.0, description="GPS тактыгы (метр)")
    device_info: Optional[str] = Field(None, max_length=255, description="Түзмөктүн маалыматы")
    device_id: Optional[str] = Field(
        None,
        min_length=3,
        max_length=255,
        description="Катталган түзмөктүн идентификатору (device binding күйгүзүлгөндө талап кылынат)",
    )



class SchoolLocalTimes(BaseModel):
    """Emits attendance timestamps in the school's own timezone.

    The columns are TIMESTAMPTZ, so a value read back from PostgreSQL arrives
    as UTC no matter what timezone it was written in. Serialized as-is it
    becomes `...Z`, and every client would then need the school's offset just
    to show a wall clock. Normalizing here keeps the documented contract:
    a time in an attendance response is the time at the school.
    """

    # Callers pass the school's zone; never part of the response body.
    school_timezone: Optional[str] = Field(default=None, exclude=True)

    @model_validator(mode="after")
    def _localize_times(self) -> "SchoolLocalTimes":
        for field in ("check_in_time", "check_out_time"):
            value = getattr(self, field, None)
            if value is not None:
                setattr(self, field, to_school_timezone(value, self.school_timezone))
        return self


class AttendanceEventRead(BaseModel):
    id: str
    teacher_id: str
    school_id: str
    event_type: AttendanceEventType
    event_time: datetime
    status: AttendanceStatus
    late_minutes: int
    distance_meters: float
    location_accuracy_meters: float
    location_verified: bool
    qr_verified: bool
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DailyAttendanceRead(SchoolLocalTimes):
    display_status: Optional[str] = None
    id: str
    teacher_id: str
    school_id: str
    date: date
    check_in_time: Optional[datetime] = None
    check_out_time: Optional[datetime] = None
    status: AttendanceStatus
    late_minutes: int
    worked_minutes: int
    is_manually_corrected: bool
    correction_reason: Optional[str] = None
    teacher_name: Optional[str] = None
    employee_code: Optional[str] = None
    phone_number: Optional[str] = None
    subject: Optional[str] = None
    lesson_delays: List[LessonDelayRead] = []
    lesson_late_minutes: int = 0
    total_late_minutes: int = 0

    model_config = ConfigDict(from_attributes=True)


class DailyAttendancePage(BaseModel):
    """Paged school-wide history, so a report is one request instead of one
    per teacher."""

    items: List[DailyAttendanceRead]
    total: int
    skip: int
    limit: int


class TodayStatusResponse(SchoolLocalTimes):
    school_name: Optional[str] = None
    date: date
    # The school's offset from UTC right now, so the client renders and
    # expires its cache against the school's day rather than guessing.
    utc_offset_minutes: int = 0
    # The school's clock at the moment of the response. A teacher deciding
    # whether they are late must not be reading their phone's clock, which
    # can be minutes off; the app ticks forward from this instead.
    server_time: Optional[datetime] = None
    display_status: Literal['ON_TIME', 'LATE', 'ABSENT', 'EXCUSED', 'DAY_OFF', 'PENDING', 'NO_SCHEDULE']
    has_checked_in: bool
    has_checked_out: bool
    check_in_time: Optional[datetime] = None
    check_out_time: Optional[datetime] = None
    status: Optional[AttendanceStatus] = None
    late_minutes: int = 0
    worked_minutes: int = 0
    scheduled_start: Optional[time] = None
    scheduled_end: Optional[time] = None
    is_day_off: bool = False
    lesson_delays: List[LessonDelayRead] = []
    lesson_late_minutes: int = 0
    total_late_minutes: int = 0


class AdminDashboardSummary(BaseModel):
    total_teachers: int
    checked_in_count: int
    on_time_count: int
    late_count: int
    not_checked_in_count: int
    date: date
    records: List[DailyAttendanceRead]


class ManualCorrectionRequest(BaseModel):
    daily_attendance_id: Optional[str] = None
    teacher_id: str
    target_date: date
    check_in_time: Optional[datetime] = None
    check_out_time: Optional[datetime] = None
    status: AttendanceStatus
    reason: str = Field(..., min_length=5, max_length=500, description="Оңдоонун себеби (сөзсүз)")


class AttendanceResetRequest(BaseModel):
    confirmation: str = Field(..., min_length=10, max_length=32)

File: packages/admin_core/lib/src/models/attendance.dart
/// A delay for a single lesson, recorded by an administrator.
class LessonDelay {
  const LessonDelay({
    required this.id,
    required this.teacherId,
    required this.schoolId,
    required this.date,
    required this.lessonNumber,
    required this.delayMinutes,
    this.reason,
    this.teacherName,
    this.createdAt,
  });

  final String id;
  final String teacherId;
  final String schoolId;
  final String date;
  final int lessonNumber;
  final int delayMinutes;
  final String? reason;
  final String? teacherName;
  final String? createdAt;

  factory LessonDelay.fromJson(Map<String, dynamic> json) {
    return LessonDelay(
      id: json['id'] as String,
      teacherId: json['teacher_id'] as String? ?? '',
      schoolId: json['school_id'] as String? ?? '',
      date: json['date'] as String? ?? '',
      lessonNumber: json['lesson_number'] as int? ?? 0,
      delayMinutes: json['delay_minutes'] as int? ?? 0,
      reason: json['reason'] as String?,
      teacherName: json['teacher_name'] as String?,
      createdAt: json['created_at'] as String?,
    );
  }
}


/// One teacher's attendance for one day, as `DailyAttendanceRead` returns it.
///
/// Times arrive already localized to the school's timezone — the API converts
/// them, because the TIMESTAMPTZ columns read back as UTC — so the wall clock
/// in the string is the one to show.
class DailyAttendance {
  const DailyAttendance({
    required this.id,
    required this.teacherId,
    required this.schoolId,
    required this.date,
    required this.status,
    this.checkInTime,
    this.checkOutTime,
    this.lateMinutes = 0,
    this.lessonLateMinutes = 0,
    this.totalLateMinutes = 0,
    this.workedMinutes = 0,
    this.isManuallyCorrected = false,
    this.correctionReason,
    this.teacherName,
    this.employeeCode,
    this.phoneNumber,
    this.subject,
    this.lessonDelays = const [],
  });

  final String id;
  final String teacherId;
  final String schoolId;

  /// `YYYY-MM-DD` in the school's timezone.
  final String date;

  /// What to show. The API's `display_status` already accounts for the
  /// schedule and the time of day, so it is preferred over the stored
  /// `status`, which is only what was recorded.
  final String status;

  final String? checkInTime;
  final String? checkOutTime;
  final int lateMinutes;
  final int lessonLateMinutes;
  final int totalLateMinutes;
  final int workedMinutes;
  final bool isManuallyCorrected;
  final String? correctionReason;
  final String? teacherName;
  final String? employeeCode;
  final String? phoneNumber;
  final String? subject;

  /// Per-lesson delays recorded for this day, if the endpoint includes them.
  final List<LessonDelay> lessonDelays;

  bool get hasCheckedIn => checkInTime != null;
  bool get hasCheckedOut => checkOutTime != null;

  factory DailyAttendance.fromJson(Map<String, dynamic> json) {
    final late = json['late_minutes'] as int? ?? 0;
    final lessonLate = json['lesson_late_minutes'] as int? ?? 0;
    return DailyAttendance(
      id: json['id'] as String,
      teacherId: json['teacher_id'] as String? ?? '',
      schoolId: json['school_id'] as String? ?? '',
      date: json['date'] as String? ?? '',
      status:
          json['display_status'] as String? ??
          json['status'] as String? ??
          'UNKNOWN',
      checkInTime: json['check_in_time'] as String?,
      checkOutTime: json['check_out_time'] as String?,
      lateMinutes: late,
      lessonLateMinutes: lessonLate,
      totalLateMinutes: json['total_late_minutes'] as int? ?? late + lessonLate,
      workedMinutes: json['worked_minutes'] as int? ?? 0,
      isManuallyCorrected: json['is_manually_corrected'] as bool? ?? false,
      correctionReason: json['correction_reason'] as String?,
      teacherName: json['teacher_name'] as String?,
      employeeCode: json['employee_code'] as String?,
      phoneNumber: json['phone_number'] as String?,
      subject: json['subject'] as String?,
      lessonDelays: (json['lesson_delays'] as List<dynamic>? ?? const [])
          .map((d) => LessonDelay.fromJson(d as Map<String, dynamic>))
          .toList(),
    );
  }
}

/// The admin dashboard for one day, as `AdminDashboardSummary` returns it.
class AttendanceDashboard {
  const AttendanceDashboard({
    required this.date,
    required this.totalTeachers,
    required this.checkedInCount,
    required this.onTimeCount,
    required this.lateCount,
    required this.notCheckedInCount,
    required this.records,
  });

  final String date;
  final int totalTeachers;
  final int checkedInCount;
  final int onTimeCount;
  final int lateCount;
  final int notCheckedInCount;
  final List<DailyAttendance> records;

  factory AttendanceDashboard.fromJson(Map<String, dynamic> json) {
    return AttendanceDashboard(
      date: json['date'] as String? ?? '',
      totalTeachers: json['total_teachers'] as int? ?? 0,
      checkedInCount: json['checked_in_count'] as int? ?? 0,
      onTimeCount: json['on_time_count'] as int? ?? 0,
      lateCount: json['late_count'] as int? ?? 0,
      notCheckedInCount: json['not_checked_in_count'] as int? ?? 0,
      records: (json['records'] as List<dynamic>? ?? const [])
          .map((r) => DailyAttendance.fromJson(r as Map<String, dynamic>))
          .toList(),
    );
  }
}

/// A page of attendance rows, as `/attendance/history` returns it.
class AttendancePage {
  const AttendancePage({required this.items, required this.total});

  final List<DailyAttendance> items;
  final int total;

  factory AttendancePage.fromJson(Map<String, dynamic> json) {
    final rows = json['items'] as List<dynamic>? ?? const [];
    return AttendancePage(
      items: rows
          .map((r) => DailyAttendance.fromJson(r as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int? ?? rows.length,
    );
  }
}
